README.md
# MD302X
MD302X is an improvement mod for MSFS default MD302 unit (aka AS1000 backup). 
Currently, it adds a one-hour battery that allows the unit to work in case of an electrical failure.

## Installation
Go to releases, download the latest release, and extract it to your Community folder.

## About
This mod is open source under the GPLv3 license. Any contributes are welcomed.

Copyright © 2021 Abdullah Radwan

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.1 - 2021-03-14
Initial release.
### Added
- One-hour battery that allows the unit to be working in case of electrical failure.

MSFS/html_ui/Pages/VCockpit/Instruments/NavSystems/AS1000_BackupDisplay/Speed/AS1000_SpeedBackup.js
class AS1000_SpeedBackup extends NavSystem {
    constructor() {
        super(...arguments);
        this.elecAvail = false;
        this.batteryTime = 3600000;
        this.shutdownTimer = 0;        
    }
    get templateID() { return "AS1000_SpeedBackup"; }
    connectedCallback() {
        super.connectedCallback();
        this.addIndependentElementContainer(new NavSystemElementContainer("Altimeter", "SvgMain", new Backup_Altimeter()));
        this.addIndependentElementContainer(new NavSystemElementContainer("Speed", "SvgMain", new Backup_Airspeed()));
    }
    Update() {
        this.updateBattery();
        super.Update();
    }
    isElectricityAvailable() {
        return this.elecAvail;
    }
    updateBattery() {
        let elecAvail;

        if (this.electricalLogic)
            elecAvail = this.electricalLogic.getValue() != 0;
        else
            elecAvail = SimVar.GetSimVarValue("CIRCUIT AVIONICS ON", "Bool");

        if (elecAvail && this.batteryTime < 3600000)
            this.batteryTime += this.deltaTime * 2;
        else if (!elecAvail && this.batteryTime > 0)
            this.batteryTime -= this.deltaTime;

        this.elecAvail = this.batteryTime > 0;

        if (this.elecAvail && Simplane.getIndicatedSpeed() < 40) {
            if (elecAvail)
                this.shutdownTimer = 60000;
            else {
                if (this.shutdownTimer <= 0)
                    this.elecAvail = false;
                else
                    this.shutdownTimer -= this.deltaTime;
            }
        }
    }
}
class Backup_Airspeed extends NavSystemElement {
    constructor() {
        super();
        this.lastIndicatedSpeed = -10000;
        this.maxSpeed = 0;
    }
    init(root) {
        this.airspeedElement = this.gps.getChildById("SvgMain");
        this.airspeedElement.setAttribute("is-backup", "true");
        var cockpitSettings = SimVar.GetGameVarValue("", "GlassCockpitSettings");
        if (cockpitSettings && cockpitSettings.AirSpeed.Initialized) {
            this.airspeedElement.setAttribute("min-speed", cockpitSettings.AirSpeed.lowLimit.toString());
            this.airspeedElement.setAttribute("green-begin", cockpitSettings.AirSpeed.greenStart.toString());
            this.airspeedElement.setAttribute("green-end", cockpitSettings.AirSpeed.greenEnd.toString());
            this.airspeedElement.setAttribute("flaps-begin", cockpitSettings.AirSpeed.whiteStart.toString());
            this.airspeedElement.setAttribute("flaps-end", cockpitSettings.AirSpeed.whiteEnd.toString());
            this.airspeedElement.setAttribute("yellow-begin", cockpitSettings.AirSpeed.yellowStart.toString());
            this.airspeedElement.setAttribute("yellow-end", cockpitSettings.AirSpeed.yellowEnd.toString());
            this.airspeedElement.setAttribute("red-begin", cockpitSettings.AirSpeed.redStart.toString());
            this.airspeedElement.setAttribute("red-end", cockpitSettings.AirSpeed.redEnd.toString());
            this.airspeedElement.setAttribute("max-speed", cockpitSettings.AirSpeed.highLimit.toString());
            this.maxSpeed = cockpitSettings.AirSpeed.highLimit;
        }
        else {
            var designSpeeds = Simplane.getDesignSpeeds();
            this.airspeedElement.setAttribute("green-begin", designSpeeds.VS1.toString());
            this.airspeedElement.setAttribute("green-end", designSpeeds.VNo.toString());
            this.airspeedElement.setAttribute("flaps-begin", designSpeeds.VS0.toString());
            this.airspeedElement.setAttribute("flaps-end", designSpeeds.VFe.toString());
            this.airspeedElement.setAttribute("yellow-begin", designSpeeds.VNo.toString());
            this.airspeedElement.setAttribute("yellow-end", designSpeeds.VNe.toString());
            this.airspeedElement.setAttribute("red-begin", designSpeeds.VNe.toString());
            this.airspeedElement.setAttribute("red-end", designSpeeds.VMax.toString());
            this.airspeedElement.setAttribute("max-speed", designSpeeds.VNe.toString());
            this.maxSpeed = designSpeeds.VNe;
        }
        if (this.gps) {
            var aspectRatio = this.gps.getAspectRatio();
            this.airspeedElement.setAttribute("aspect-ratio", aspectRatio.toString());
        }
    }
    onEnter() {
    }
    isReady() {
        return true;
    }
    onUpdate(_deltaTime) {
        var indicatedSpeed = Simplane.getIndicatedSpeed();
        if (indicatedSpeed != this.lastIndicatedSpeed) {
            this.airspeedElement.setAttribute("airspeed", indicatedSpeed.toFixed(1));
            this.lastIndicatedSpeed = indicatedSpeed;
        }
        let crossSpeed = SimVar.GetGameVarValue("AIRCRAFT CROSSOVER SPEED", "Knots");
        let cruiseMach = SimVar.GetGameVarValue("AIRCRAFT CRUISE MACH", "mach");
        let crossSpeedFactor = Simplane.getCrossoverSpeedFactor(this.maxSpeed, cruiseMach);
        if (crossSpeed != 0) {
            this.airspeedElement.setAttribute("max-speed", (Math.min(crossSpeedFactor, 1) * this.maxSpeed).toString());
        }
    }
    onExit() {
    }
    onEvent(_event) {
    }
}
class Backup_Altimeter extends NavSystemElement {
    constructor() {
        super();
        this.lastAltitude = -10000;
        this.lastPressure = -10000;
    }
    init(root) {
        this.altimeterElement = this.gps.getChildById("SvgMain");
        this.altimeterElement.setAttribute("is-backup", "true");
        if (this.gps) {
            var aspectRatio = this.gps.getAspectRatio();
            this.altimeterElement.setAttribute("aspect-ratio", aspectRatio.toString());
        }
    }
    onEnter() {
    }
    isReady() {
        return true;
    }
    onUpdate(_deltaTime) {
        var altitude = SimVar.GetSimVarValue("INDICATED ALTITUDE:2", "feet");
        if (altitude != this.lastAltitude) {
            this.lastAltitude = altitude;
            this.altimeterElement.setAttribute("altitude", altitude);
        }
        var pressure = SimVar.GetSimVarValue("KOHLSMAN SETTING HG:2", "inches of mercury");
        pressure = fastToFixed(pressure, 2);
        if (pressure != this.lastPressure) {
            this.lastPressure = pressure;
            this.altimeterElement.setAttribute("pressure", pressure);
        }
    }
    onExit() {
    }
    onEvent(_event) {
        switch (_event) {
            case "BARO_INC":
                SimVar.SetSimVarValue("K:KOHLSMAN_INC", "number", 1);
                break;
            case "BARO_DEC":
                SimVar.SetSimVarValue("K:KOHLSMAN_DEC", "number", 1);
                break;
        }
    }
}
registerInstrument("as1000-speedbackup-element", AS1000_SpeedBackup);
//# sourceMappingURL=AS1000_SpeedBackup.js.map

MSFS/html_ui/Pages/VCockpit/Instruments/NavSystems/AS1000_BackupDisplay/Attitude/AS1000_AttitudeBackup.js
class AS1000_AttitudeBackup extends NavSystem {
    constructor() {
        super(...arguments);
        this.elecAvail = false;
        this.batteryTime = 3600000;
        this.shutdownTimer = 0;        
    }
    get templateID() { return "AS1000_AttitudeBackup"; }
    connectedCallback() {
        super.connectedCallback();
        this.addIndependentElementContainer(new NavSystemElementContainer("Horizon", "Horizon", new Backup_Attitude()));
    }
    Update() {
        this.updateBattery();
        super.Update();
    }
    isElectricityAvailable() {
        return this.elecAvail;
    }
    updateBattery() {
        let elecAvail;

        if (this.electricalLogic)
            elecAvail = this.electricalLogic.getValue() != 0;
        else
            elecAvail = SimVar.GetSimVarValue("CIRCUIT AVIONICS ON", "Bool");

        if (elecAvail && this.batteryTime < 3600000)
            this.batteryTime += this.deltaTime * 2;
        else if (!elecAvail && this.batteryTime > 0)
            this.batteryTime -= this.deltaTime;

        this.elecAvail = this.batteryTime > 0;

        if (this.elecAvail && Simplane.getIndicatedSpeed() < 40) {
            if (elecAvail)
                this.shutdownTimer = 60000;
            else {
                if (this.shutdownTimer <= 0)
                    this.elecAvail = false;
                else
                    this.shutdownTimer -= this.deltaTime;
            }
        }
    }
    IsGlassCockpit() { return true; }
}
class Backup_Attitude extends NavSystemElement {
    constructor() {
        super();
        this.vDir = new Vec2();
    }
    init(root) {
        this.attitudeElement = this.gps.getChildById("Horizon");
        this.attitudeElement.setAttribute("is-backup", "true");
        if (this.gps) {
            var aspectRatio = this.gps.getAspectRatio();
            this.attitudeElement.setAttribute("aspect-ratio", aspectRatio.toString());
        }
    }
    onEnter() {
    }
    onUpdate(_deltaTime) {
        var xyz = Simplane.getOrientationAxis();
        if (xyz) {
            this.attitudeElement.setAttribute("pitch", (xyz.pitch / Math.PI * 180).toString());
            this.attitudeElement.setAttribute("bank", (xyz.bank / Math.PI * 180).toString());
            this.attitudeElement.setAttribute("slip_skid", Simplane.getInclinometer().toString());
        }
    }
    onExit() {
    }
    onEvent(_event) {
    }
}
registerInstrument("as1000-attitudebackup-element", AS1000_AttitudeBackup);
//# sourceMappingURL=AS1000_AttitudeBackup.js.map

MSFS/html_ui/Pages/VCockpit/Instruments/NavSystems/AS1000_BackupDisplay/AttitudeSpeed/AS1000_AttitudeSpeedBackup.js
class AS1000_AttitudeSpeedBackup extends NavSystem {
    constructor() {
        super(...arguments);
        this.altimeterIndex = 2;
        this.elecAvail = false;
        this.batteryTime = 3600000;
        this.shutdownTimer = 0;
    }
    get templateID() { return "AS1000_AttitudeSpeedBackup"; }
    connectedCallback() {
        super.connectedCallback();
        this.addIndependentElementContainer(new NavSystemElementContainer("Altimeter", "SvgMain", new ASBackup_Altimeter()));
        this.addIndependentElementContainer(new NavSystemElementContainer("Speed", "SvgMain", new ASBackup_Airspeed()));
        this.addIndependentElementContainer(new NavSystemElementContainer("Horizon", "Horizon", new ASBackup_Attitude()));
    }
    parseXMLConfig() {
        super.parseXMLConfig();
        if (this.instrumentXmlConfig) {
            let altimeterIndexElems = this.instrumentXmlConfig.getElementsByTagName("AltimeterIndex");
            if (altimeterIndexElems.length > 0) {
                this.altimeterIndex = parseInt(altimeterIndexElems[0].textContent) + 1;
            }
        }
    }
    Update() {
        this.updateBattery();
        super.Update();
    }
    isElectricityAvailable() {
        return this.elecAvail;
    }
    updateBattery() {
        let elecAvail;

        if (this.electricalLogic)
            elecAvail = this.electricalLogic.getValue() != 0;
        else
            elecAvail = SimVar.GetSimVarValue("CIRCUIT AVIONICS ON", "Bool");

        if (elecAvail && this.batteryTime < 3600000)
            this.batteryTime += this.deltaTime * 2;
        else if (!elecAvail && this.batteryTime > 0)
            this.batteryTime -= this.deltaTime;

        this.elecAvail = this.batteryTime > 0;

        if (this.elecAvail && Simplane.getIndicatedSpeed() < 40) {
            if (elecAvail)
                this.shutdownTimer = 60000;
            else {
                if (this.shutdownTimer <= 0)
                    this.elecAvail = false;
                else
                    this.shutdownTimer -= this.deltaTime;
            }
        }
    }
}
class ASBackup_Airspeed extends NavSystemElement {
    constructor() {
        super();
        this.lastIndicatedSpeed = -10000;
    }
    init(root) {
        this.airspeedElement = this.gps.getChildById("SvgMain");
        this.airspeedElement.setAttribute("is-backup", "true");
        var cockpitSettings = SimVar.GetGameVarValue("", "GlassCockpitSettings");
        if (cockpitSettings && cockpitSettings.AirSpeed.Initialized) {
            this.airspeedElement.setAttribute("min-speed", cockpitSettings.AirSpeed.lowLimit.toString());
            this.airspeedElement.setAttribute("green-begin", cockpitSettings.AirSpeed.greenStart.toString());
            this.airspeedElement.setAttribute("green-end", cockpitSettings.AirSpeed.greenEnd.toString());
            this.airspeedElement.setAttribute("flaps-begin", cockpitSettings.AirSpeed.whiteStart.toString());
            this.airspeedElement.setAttribute("flaps-end", cockpitSettings.AirSpeed.whiteEnd.toString());
            this.airspeedElement.setAttribute("yellow-begin", cockpitSettings.AirSpeed.yellowStart.toString());
            this.airspeedElement.setAttribute("yellow-end", cockpitSettings.AirSpeed.yellowEnd.toString());
            this.airspeedElement.setAttribute("red-begin", cockpitSettings.AirSpeed.redStart.toString());
            this.airspeedElement.setAttribute("red-end", cockpitSettings.AirSpeed.redEnd.toString());
            this.airspeedElement.setAttribute("max-speed", cockpitSettings.AirSpeed.highLimit.toString());
        }
        else {
            var designSpeeds = Simplane.getDesignSpeeds();
            this.airspeedElement.setAttribute("green-begin", designSpeeds.VS1.toString());
            this.airspeedElement.setAttribute("green-end", designSpeeds.VNo.toString());
            this.airspeedElement.setAttribute("flaps-begin", designSpeeds.VS0.toString());
            this.airspeedElement.setAttribute("flaps-end", designSpeeds.VFe.toString());
            this.airspeedElement.setAttribute("yellow-begin", designSpeeds.VNo.toString());
            this.airspeedElement.setAttribute("yellow-end", designSpeeds.VNe.toString());
            this.airspeedElement.setAttribute("red-begin", designSpeeds.VNe.toString());
            this.airspeedElement.setAttribute("red-end", designSpeeds.VMax.toString());
            this.airspeedElement.setAttribute("max-speed", designSpeeds.VNe.toString());
        }
        if (this.gps) {
            var aspectRatio = this.gps.getAspectRatio();
            this.airspeedElement.setAttribute("aspect-ratio", aspectRatio.toString());
        }
    }
    onEnter() {
    }
    isReady() {
        return true;
    }
    onUpdate(_deltaTime) {
        var indicatedSpeed = Simplane.getIndicatedSpeed();
        if (indicatedSpeed != this.lastIndicatedSpeed) {
            this.airspeedElement.setAttribute("airspeed", indicatedSpeed.toFixed(1));
            this.lastIndicatedSpeed = indicatedSpeed;
        }
    }
    onExit() {
    }
    onEvent(_event) {
    }
}
class ASBackup_Altimeter extends NavSystemElement {
    constructor() {
        super();
        this.lastAltitude = -10000;
        this.lastPressure = -10000;
    }
    init(root) {
        this.altimeterElement = this.gps.getChildById("SvgMain");
        this.altimeterElement.setAttribute("is-backup", "true");
        if (this.gps) {
            var aspectRatio = this.gps.getAspectRatio();
            this.altimeterElement.setAttribute("aspect-ratio", aspectRatio.toString());
        }
    }
    onEnter() {
    }
    isReady() {
        return true;
        ;
    }
    onUpdate(_deltaTime) {
        var altitude = SimVar.GetSimVarValue("INDICATED ALTITUDE:" + this.gps.altimeterIndex, "feet");
        if (altitude != this.lastAltitude) {
            this.lastAltitude = altitude;
            this.altimeterElement.setAttribute("altitude", altitude);
        }
        var pressure = SimVar.GetSimVarValue("KOHLSMAN SETTING HG:" + this.gps.altimeterIndex, "inches of mercury");
        pressure = fastToFixed(pressure, 2);
        if (pressure != this.lastPressure) {
            this.lastPressure = pressure;
            this.altimeterElement.setAttribute("pressure", pressure);
        }
    }
    onExit() {
    }
    onEvent(_event) {
        switch (_event) {
            case "BARO_INC":
                SimVar.SetSimVarValue("K:KOHLSMAN_INC", "number", this.gps.altimeterIndex);
                break;
            case "BARO_DEC":
                SimVar.SetSimVarValue("K:KOHLSMAN_DEC", "number", this.gps.altimeterIndex);
                break;
        }
    }
}
class ASBackup_Attitude extends NavSystemElement {
    constructor() {
        super(...arguments);
        this.vDir = new Vec2();
    }
    init(root) {
        this.attitudeElement = this.gps.getChildById("Horizon");
        this.attitudeElement.setAttribute("is-backup", "true");
        if (this.gps) {
            var aspectRatio = this.gps.getAspectRatio();
            this.attitudeElement.setAttribute("aspect-ratio", aspectRatio.toString());
        }
    }
    onEnter() {
    }
    onUpdate(_deltaTime) {
        var xyz = Simplane.getOrientationAxis();
        if (xyz) {
            this.attitudeElement.setAttribute("pitch", (xyz.pitch / Math.PI * 180).toString());
            this.attitudeElement.setAttribute("bank", (xyz.bank / Math.PI * 180).toString());
            this.attitudeElement.setAttribute("slip_skid", Simplane.getInclinometer().toString());
        }
    }
    onExit() {
    }
    onEvent(_event) {
    }
}
registerInstrument("as1000-attitudespeedbackup-element", AS1000_AttitudeSpeedBackup);
//# sourceMappingURL=AS1000_AttitudeSpeedBackup.js.map